cpp/server/handler_v2.cc

```diff
@@ -0,0 +1,185 @@
+  // automatically mean we're unable to answer this request.
+  if (IsNodeStale()) {
+    // Can't do this on the libevent thread since it can block on the lock in
+    // ClusterStatusController::GetFreshNodes().
+    pool_->Add(bind(&Proxy::ProxyRequest, proxy_, request));
+  } else {
+    local_handler(request);
+  }
+}
+
+
+void HttpHandlerV2::AddProxyWrappedHandler(
+    libevent::HttpServer* server, const string& path,
+    const libevent::HttpServer::HandlerCallback& local_handler) {
+  const libevent::HttpServer::HandlerCallback stats_handler(
+      bind(&StatsHandlerInterceptor, path, local_handler, _1));
+  CHECK(server->AddHandler(path, bind(&HttpHandlerV2::ProxyInterceptor, this,
+                                      stats_handler, _1)));
+}
+
+
+void HttpHandlerV2::Add(libevent::HttpServer* server) {
+  CHECK_NOTNULL(server);
+  // TODO(pphaneuf): An optional prefix might be nice?
+  // TODO(pphaneuf): Find out which methods are CPU intensive enough
+  // that they should be spun off to the thread pool.
+  AddProxyWrappedHandler(server, "/ct/v2/get-entries",
+                         bind(&HttpHandlerV2::GetEntries, this, _1));
+  AddProxyWrappedHandler(server, "/ct/v2/get-proof-by-hash",
+                         bind(&HttpHandlerV2::GetProof, this, _1));
+  AddProxyWrappedHandler(server, "/ct/v2/get-sth",
+                         bind(&HttpHandlerV2::GetSTH, this, _1));
+  AddProxyWrappedHandler(server, "/ct/v2/get-sth-consistency",
+                         bind(&HttpHandlerV2::GetConsistency, this, _1));
+
+  // Now add any sub-class handlers.
+  AddHandlers(server);
+}
+
+
+void HttpHandlerV2::SetProxy(Proxy* proxy) {
+  LOG_IF(FATAL, proxy_) << "Attempting to re-add a HttpHandlerV2.";
+  proxy_ = CHECK_NOTNULL(proxy);
+}
+
+
+void HttpHandlerV2::GetEntries(evhttp_request* req) const {
+  return SendJsonError(event_base_, req, HTTP_NOTIMPLEMENTED,
+                       "Not yet implemented.");
+}
+
+
+void HttpHandlerV2::GetProof(evhttp_request* req) const {
+  return SendJsonError(event_base_, req, HTTP_NOTIMPLEMENTED,
+                       "Not yet implemented.");
+}
+
+
+void HttpHandlerV2::GetSTH(evhttp_request* req) const {
+  return SendJsonError(event_base_, req, HTTP_NOTIMPLEMENTED,
+                       "Not yet implemented.");
+}
+
+
+void HttpHandlerV2::GetConsistency(evhttp_request* req) const {
+  return SendJsonError(event_base_, req, HTTP_NOTIMPLEMENTED,
+                       "Not yet implemented.");
+}
+
+
+void HttpHandlerV2::BlockingGetEntries(evhttp_request* req, int64_t start,
+                                     int64_t end, bool include_scts) const {
+  return SendJsonError(event_base_, req, HTTP_NOTIMPLEMENTED,
+                       "Not yet implemented.");
+}
+
+
+bool HttpHandlerV2::IsNodeStale() const {
+  return staleness_tracker_->IsNodeStale();
+}
+
+
+void HttpHandlerV2::UpdateNodeStaleness() {
+  staleness_tracker_->UpdateNodeStaleness();
```

Comment:
When I embark on a refactoring, I often do a "pre-refactoring" of trimming off any dead code. Non-existent code being the easiest to refactor... :wink: